### ISSUE TITLE
DelayedCall should use unsignedID in str method

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3791]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3791|
|Type|enhancement|
|Created|2009-04-22 08:38:20Z|
|Branch|https://github.com/twisted/twisted/tree/unsigned-id-delayed-call-3791|

Really.

Attachments:

* [unsignedID.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/278/27854e15b10dd3a6a346df1693666731c5a53cfc/7b17aed5c9eabb6f1c75f5a9fe49dc0e6cebaa31.diff) (3336 bytes) - added by drake on 2009-05-02 20:59:54Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3791 3791
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_unsigned_id_delayed_call_3791 branches/unsigned-id-delayed-call-3791
branch_author__drake__itamar__exarkun drake, itamar, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1240389500000000 1240389500000000
changetime__1243883338000000 1243883338000000
version__None None
owner__ 
cc__drake
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="drake's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|drake removed owner|
|-|-|

The attached patch addresses the ticket and adds a test that verifies that unsignedID is used.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [26901]) Branching to 'unsigned-id-delayed-call-3791'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26902]) Apply unsignedID.diff

refs [#3791](https://github.com/twisted/twisted/issues/3791)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [26903]) This isn't really necessary to implement this change

refs [#3791](https://github.com/twisted/twisted/issues/3791)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `new`|
|-|-|

I applied the diff to a branch so I can muck about with it a bit:

  * Making `DelayedCall` new-style isn't necessary to resolve this ticket, and it's a big change I'd rather not think about now. :)  I removed it.
  * I looked around for prior art wrt `unsignedID` testing.  There's a test for `unsignedID` itself and an AMP test which exercises `unsignedID`-using but doesn't directly test the use of `unsignedID`.  I think we should make testing for this kind of thing nice and easy, since presumably every `__repr__` we have should be testing this.  So I'm going to give that a go.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun removed owner|
|-|-|

Okay...  I introduced `setIDFunction` and used it in some tests.  Maybe it's a lame idea.  Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

Looks fine to me, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [26967]) Merge unsigned-id-delayed-call-3791

Author: drake, exarkun
Reviewer: exarkun, therve
Fixes: [#3791](https://github.com/twisted/twisted/issues/3791)

Add `twisted.python.util.setIDFunction` to allow the behavior of `unsignedID`
to be controlled.  Then use this to test that `unsignedID` is being used in
`DelayedCall` and `AMP` to construct the repr string.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
